### ISSUE TITLE
Allow extensions to abort reverse fires

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#98]
In https://github.com/WordPoints/hooks-api/issues/97#issuecomment-183680925 we realized that up to now we've been arbitrarily dictating that all reverse fires are hits. Whether the reactor actually treats them as such has been up to it to decide, but the reverse firer has always called the reversal method on the reactor without giving extensions any opportunity to protest. The idea to change this first came to me while working on https://github.com/WordPoints/hooks-api/issues/66#issuecomment-183487741, and I think we ought to do it.

The main concern that has been lingering in my mind is that a reversal miss would put the entity status out of sync with the reaction. This is a concern only for the toggle events, whose expected behavior (by the points reactor) is predicated on the idea that no rehits can occur—the reaction can't hit twice in a row for a particular entity without a reversal in between. But if a reverse fire of the event can occur without the hit actually being reversed, this assumption is no longer true.

Of course, the converse is already true: the entity can toggle to the "on" position, firing the event, without a hit being recorded, because the extensions can block it. The difference is that this doesn't actually have the effect of multiple reverses firing without hits in between them. This is because the reverse firer only actually fires for the actual un-superseded hits from the logs (#68). So reversals aren't just opposites of regular fires, but actually reverse specific hits by those fires.

The key here is that the "expected behavior" we're talking about is just that: the behavior that one particular reactor is trying to achieve using the events. The behavior isn't by very nature of the events themselves or the firers (something discussed in #97 too). So what is expected will vary form reactor to reactor, and possibly even from reaction to reaction. It is up to the developer to properly code the reactor and any custom extensions necessary to achieve the "expected behavior" for a particular use-case.

In other words, there are times, as is the general case with the points reactor, that we never want doubled-up hits to occur. But there is no reason that another reactor, even though it performs some reverses, might not want to skip others, and expect hits to jut go on normally, without being arbitrarily blocked in some way.

And so there is no reason that we should not give extensions the ability to abort reverse fires. This has the potential to greatly increase the usefulness of the API, essentially making it twice as flexible.

**Comment by JDGrimes** [comment]
## Only Latest Hits Reversed?

The only thing that we might want to consider is if the behavior displayed by the reverse firer should change. It currently pulls all un-superseded hits from the database when the reverse fire is triggered. But if the reversal of one of them has been aborted, shouldn't that one be skipped over again, and not even subjected to the same treatment? In other words, I think that only the latest hit for an entity for each reaction should be reversed. That would be for a new ticket though.

**Comment by JDGrimes** [comment]
## Should Reactions Be Able To Affect Reversals?

In some ways this ticket seems to directly contradict the conclusion of #70. There we decided that reversals for should still fire for a hit even if the reaction was deleted. This has since been questioned, notably in https://github.com/WordPoints/hooks-api/issues/68#issuecomment-176388076, but the validity of the basic premise hasn't been overturned yet.

In essence, the argument there was that we don't perform a reversal based on the current state of the reaction (its current settings). It isn't a reverse-hit that does the opposite of what the reaction would now do, but a reversal of the previous hit, undoing whatever that hit did.

And yet now we are thinking of letting extensions abort reverses, which means that it will be possible for them to take the current state of the reactor into account.

Once again though, in that discussion the behavior expected by the points reactor seems to be confused with the nature of the reverse firer itself. Just because the points reactor chooses to handle the reverse hit in this way, doesn't mean that it is the only "right" way to do it and every other way is "wrong". There is nothing to prevent another reactor from wanting to take the current settings of the reaction into account. It's all just what the reactor chooses to do. We even admitted as much in #70, pointing out that reactors could choose to ignore reversals for deleted reactions if they wanted to.

### What is the reverse firer's job?

So on the one hand the expected behavior depends upon what the developer is trying to achieve with a particular reactor. However, this doesn't entirely exempt the firer from restricting its own behavior based on what it is by nature trying to do. It might be that the reactor needs to listen to several different firers to achieve its desired ends.

I think perhaps that this is what we have a case of here. The reverse firer may not need to expand itself to take in every possible reversal use-case, but should stay true to its nature. The other behaviors could then be supplied by other firers.

The big question of course is what the nature of the reverse firer is, or should be. I think that #70 may have actually pegged this pretty well, by saying that a reverse fire is expected to result in a reversal of a previous hit. This is the premise that #68 is based upon, and the reason why we query the hit logs to find out what is being reversed. (The real remaining problem there might just be that we've tied a reaction too closely to the definition of a fire.)

### What should a reversal be?

The big question though, is what reversing that hit should mean. Should it have to mean undoing it, or should it be able to mean doing something else more or less contradictory to the original hit, but not necessarily returning things to an "at rest" state?

This again seems like something that it should be up to the reactor to decide. However, when we start thinking about reversing only the latest hits, we are talking about adding another restriction to the firer itself. It seems odd because the firer is catering specifically to a certain type of event (toggle events) and a certain expectation for them. The question that we have to pose is what is the meaning of the reverse action being called? How do we know which hits it is reversing?

This is the real crux of the matter, and the truth is that the answer is arbitrary and up to the firer to decide. What hits a reverse is tied to is something determined entirely by the nature of the firer.

#82 is very much in this vein, and it explains how that repeatable events cannot use the reverse firer, simply because of the very nature of it. The reverse firer has been designed to reverse a hit when a toggle event is toggled.

### What should the nature of the reverse firer be?

Having settled the (current) nature of the reverse fire, we may still question whether this is the nature that it should have, whether it can accomplish this while taking a broader approach, and if not whether perhaps some restrictions should be left up to extensions instead.

So how could we possibly make the behavior of the reverse-firer "broader"? I suppose that the broadest method of determining which hits to reverse would be to just reverse all of them. This is in fact what the reverse firer actually does at present, as noted above. Well, actually, what I mean by "all" is all un-reversed hits, not all hits including the ones that may have already been reversed. I suppose that would really be the broadest that it could be.

So I suppose we have to ask why it shouldn't be that broad? Could it still be useful, or would that destroy its usefulness as regards to toggle events? And would it be possible to add back those constraints as needed via extensions instead? Where does the job of the firer end and the job of the extensions begin?

### What is a firer?

What we're sort of trying to define here is what a firer actually is.

We could say that firers should only do that which extensions cannot do. But then we could probably argue that extensions can really do everything, so we only need a single firer. But that would require reactors to have a single method that handled all sorts of different types of hits. But that isn't what we want, the reactors shouldn't have multi-use methods, but methods that just do what they're supposed to do. But yet reactor expectations shouldn't exactly be determining firer behavior. But what should?

I think we need to step back and approach this from the other end. Let the actions and events define the nature of the firers. Then the reactors just hook into this.

### What are reversal actions?

So what are reversal actions?

Reversals aren't just an "on demand" thing. They only occur when the even that caused the hit is reversed. So what the reversal is is actually not determined by the firer, but by how the actions for that event are registered. What is still has to be determined at that point is which hits are being reversed. Or is it? Because the reverse actions have been restricted only to toggle events, which only toggle. And so the nature of these events actually dictates the way this firer handles them. It really is a firer created for a particular type of event, and rightly so: because what sort of actions relate to an event determine what sort of event it is and therefore what ways it can be fired.

So our goal is to work within the constraints set by that event type/action type/expectation for the fire type. And ideally, we would work exactly at the limit of those constraints, with any additional constraints being provided by extensions.

In other words, the expectations for a firer are set by the events, not the reactors.

The toggle events are events that occur when the status of an entity is toggled. The event cannot have an intermediate state, it either has one or another. Actually, there can be many different status, but it can only have one at a time. We might call these quantum events because they can only occur in discrete quantum leap-like changes occur in an entity.

Anyway, the point is that the events flip back and forth between an "on" state and one or more "off" states. We don't really worry about which "off" state the entity is in, we just define the reverse as the transition away from the "on" state to something else (anything else, really).

### Conclusion

So when a reverse occurs, it is reversing a particular toggle "on" event, not all "on" events that have ever occurred.

Normally we wouldn't have to worry about this, because any "on" events would always be reversed. Each time a reversal occurred. But if we allow for extensions to prevent a reverse fire from hitting (that is, from reversing the hit), that no longer holds true, and requires special handling from the reactor, to maintain the correct behavior. And so the reversal of only the latest hits (the hits from the last toggle "on") becomes necessary.

---

Note that our new definition of the reverse firer is a departure from the way that reversals were previously viewed, as illustrated by https://github.com/WordPoints/hooks-api/issues/79#issuecomment-174195466:

> In fact, I suppose that reversals should perhaps be allowed to happen even when the initial arg doesn't have its state changed, because it is the event itself being marked as spam, not the entity. 

Of course, this is still true, it is possible for a hit to be reversed or marked as spam without the reverse event action being triggered, but the reverse firer is not a tool that can do that.
